databricks/sdk/mixins/files.py

```diff
@@ -713,11 +727,560 @@ def download(self, file_path: str) -> DownloadResponse:
-        if_unmodified_since_timestamp: Optional[str] = None,
+        target_path: str,
+        input_stream: BinaryIO,
+        session_token: str,
+        pre_read_buffer: bytes,
+        cloud_provider_session: requests.Session,
+    ):
+        current_part_number = 1
+        etags: dict = {}
+
+        # Why are we buffering the current chunk?
+        # AWS and Azure don't support traditional "Transfer-encoding: chunked", so we must
+        # provide each chunk size up front. In case of a non-seekable input stream we need
+        # to buffer a chunk before uploading to know its size. This also allows us to rewind
+        # the stream before retrying on request failure.
+        # AWS signed chunked upload: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html
+        # https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blobs-tune-upload-download-python#buffering-during-uploads
+
+        chunk_offset = 0  # used only for logging
+
+        # This buffer is expected to contain at least multipart_upload_chunk_size bytes.
+        # Note that initially buffer can be bigger (from pre_read_buffer).
+        buffer = pre_read_buffer
+
+        def fill_buffer():
+            bytes_to_read = max(0, self._config.multipart_upload_chunk_size - len(buffer))
+            if bytes_to_read > 0:
+                next_buf = input_stream.read(bytes_to_read)
+                new_buffer = buffer + next_buf
+                return new_buffer
+            else:
+                # we have already buffered enough data
+                return buffer
+
+        retry_count = 0
+        eof = False
+        while not eof:
+            # If needed, buffer the next chunk.
+            buffer = fill_buffer()
+            if not len(buffer):
+                # End of stream, no need to request the next block of upload URLs.
+                break
+
+            _LOG.debug(
+                f"Multipart upload: requesting next {self._config.multipart_upload_batch_url_count} upload URLs starting from part {current_part_number}"
+            )
+
+            body: dict = {
+                "path": target_path,
+                "session_token": session_token,
+                "start_part_number": current_part_number,
+                "count": self._config.multipart_upload_batch_url_count,
+                "expire_time": self._get_url_expire_time(),
+            }
+
+            headers = {"Content-Type": "application/json"}
+
+            # Requesting URLs for the same set of parts is an idempotent operation, safe to retry.
+            # _api.do() does retry
+            upload_part_urls_response = self._api.do(
+                "POST", "/api/2.0/fs/create-upload-part-urls", headers=headers, body=body
+            )
+            # no need to check response status, _api.do() will throw exception on failure
+
+            upload_part_urls = upload_part_urls_response.get("upload_part_urls", [])
+            if not len(upload_part_urls):
```

Comment:
Let's aim for clarity for developers who are not use to Python idioms.

```suggestion
            if len(upload_part_urls) == 0:
```